rfc-143-split-database-instances.md

```diff
@@ -0,0 +1,66 @@
+
+5. We still host some database features ourselves (such as Mongo 2.6) which has a maintenance burden and allows us to run very out of date versions of these services. 
+
+This is coming up now because our versions of MySQL and Postgres are approaching End of Life (notice for [MySQL](https://forums.aws.amazon.com/ann.jspa?annID=8790) and [Postgres](https://forums.aws.amazon.com/ann.jspa?annID=8499)), forcing us to plan major upgrades for both over the next few months.
+
+## Proposal
+
+We propose a new principle to be applied across GOV.UK:
+
+1. Supporting services for an application should not be shared with any other applications at the infrastructure/provisioning level.
+
+> Note: this principle doesn't cover services used for application communication, e.g. where SNS/SQS/RabbitMQ broadcasts events for other applications. The use of the service carries most weight here – e.g. a fictional RabbitMQ queue used solely to manage Sidekiq jobs would be considered a "supporting service", whereas a RabbitMQ queue to broadcast publishing domain events would be accessed by many apps. Both these queues would live in distinct RabbitMQ instances/servers.
+
+Applying this initially to our databases, we propose that:
+
+1. All new long-lived databases are created in their own individual RDS instances.
+2. Existing MySQL and Postgres databases are migrated out of the central instance to individual managed-service instances as part of the upcoming major upgrade cycle.
+3. Existing Redis and Mongo databases are migrated out of their central instances in-line with team priorities, and prior to any major upgrade.
+4. Any future supporting services for applications should be distinct to the application and, where applicable, run on a managed service
+
+
+### Consequences
+
+- It will be easier to identify the causes of performance problems in an application.
+
+- The chance of a problem with one service cascading to other services will be reduced, because there is no contention for resources which aren't shared.
+
+- Creating more RDS instances means that application-owning teams will need to monitor and consider database upgrades as part of their day-to-day work. We can ease this burden through monitoring & reporting, which may be considered by a future Platform Reliability team.
+
+- There will be less consistency in database versions across the programme. Individual teams are responsible for their upgrade appetite and prioritisation.
+
+- There is a potential increase in costs caused by having dedicated instances, particularly when they are initially split (because it's harded to accurately estimate resource requirements). This may be offset by being able to optimise individual databases to their needs.
```

Comment:
Do we have a rough idea of what this isolation will cost?

The consequences of the principle (redis per app, memcached per app, RDS per app, etc) may lead to a very large increase in AWS costs and underutilisation of AWS resources.